_posts/2024-04-20-coroutines.md

```diff
@@ -0,0 +1,348 @@
+
+For example, assume that we have a library for network communication that provides us coroutine to do the io with:
+
+```cpp
+network_coroutine send_data(tcp_connection& con)
+{
+    co_await con.make_data_send_awaiter("How are you?");
+    std::string response = co_await con.make_receive_data_awaiter();
+    if(response == "good")
+    {
+        co_await con.make_data_send_awaiter("Good, me too!");
+    }
+    else
+    {
+        co_await con.make_data_send_awaiter("Oh, I see");
+    }
+}
+```
+
+It is good to summarize some properties of this coroutine:
+ - It does not have to block, once the `co_await` call the parent will get control of the program flow and can just send the data provided by `awaiter` asynchronously.
+ - After that, the parent is free to ask another coroutine for data or to destroy this coroutine entirely. This makes sense mostly in cases the request to get data would fail.
+
+### More
+
+This was a fast and simple explanation of coroutines, the purpose of this article is not to give a detailed explanation. I can suggest a blog post such as this one: https://www.scs.stanford.edu/~dm/blog/c++-coroutines.html
+
+## Embedded coroutines
+
+Let's get back to embedded and try to talk about how are coroutines relevant for embedded development.
+That means possible problems with coroutines in embedded development, the relevance of coroutines, and some hints for proper usage.
+
+### Relevance
+
+A common task for embedded is to do plenty of IO communication via peripherals and in the meantime to take care of multiple processes in parallel on a single-core system.
+
+That is, we have to handle:
+ - complex computations (Matrix computations)
+ - IO (send/receive data over UART/I2C/...)
+ - long-term processes (PID regulator regulating the temperature of the room)
+
+To handle all those processes at a single point in time, we have two major approaches:
+ 1) implement it as an iterative process and schedule it
+ 2) use threads for the processes
+
+Let's take communication over UART as an example:
+ 1) iterative process means the usage of structure to store the state of the communication, and provide a function that based on the current state advances the state. (Commonly with the usage of an enum representing actual status)
+ 2) implement the communication in a thread which might release its time if it waits for some operation
+
+Approach 1) has multiple potential issues, it might take a lot of code to implement a complex exchange of data in this way (a lot of state variables) and the approach is problematic, as there is a chance that one of the steps might take longer than expected and we can't prevent that.
+
+Approach 2) has another set of issues. Each thread requires its own stack space (which might not scale), and we got all the problems of parallelism - exchanging data between can suffer various potential concurrency issues.
```

Comment:
Eeeeeh, it does, point of coroutines is that you make a context swithc at specific points in the code that are controlled by you. While in normal freertos-like threading, the context switch can happen anytime the main timer fires and switches the threads.

This means that you do not have to think problems about stuff like strict atomic work with shared variables, let's assume that i is shared variable between parts of the system that execute concurrently. This is unsafe in threads, but is perfectly ok in case of coroutines:

```cpp
int j = i;
j += 1;
i = j;
```

That is, with threads: context switch can happen at any point in time, at any instruction, and you DO have to think about it and take care
with coros: context switch happens ONLY when you explicitly write in the code that the coroutine should interrupt itself.

For me this is one of the biggest motivations to use coroutines over threads in general, as the mental overhead of managing concurrent access to stuff with this explicit cooperation is much smaller than with threads.

Anyway, I will try to re-think this part and think how to express it in better way :)